File: Java/Graphs/21_sentenceSimilarity2.java
public class Solution {
    /**
     * @param words1: 
     * @param words2: 
     * @param pairs: 
     * @return: Whether sentences are similary or not?
     */
    public boolean isSimilarDFS(HashMap<String,ArrayList<String>> graph,String src,String dest, HashSet<String> vis){
        if(src.equals(dest)) return true;
        vis.add(src);
        if(graph.get(src) == null) return false;
        ArrayList<String> al = graph.get(src);
        for(int i=0;i<al.size();i++){
            String s = al.get(i);
            if(!vis.contains(s)){
                if(isSimilarDFS(graph,s,dest,vis))  return true;
            }
        }
        return false;
    }
    public boolean areSentencesSimilarTwo(List<String> words1, List<String> words2, List<List<String>> pairs) {
        int n = words1.size();
        int m = words2.size();
        if(n != m) return false;
        HashMap<String,ArrayList<String>> graph = new HashMap<>();
        for(List<String> p: pairs){
            String u = p.get(0);
            String v = p.get(1);
            if(!graph.containsKey(u)) graph.put(u,new ArrayList<>());
            if(!graph.containsKey(v)) graph.put(v,new ArrayList<>());
            graph.get(u).add(v);
            graph.get(v).add(u);
        } 
        
        for(int i=0;i<n;i++){
            HashSet<String> vis = new HashSet<>();
            String src = words1.get(i);
            String dest = words2.get(i);
            if(!isSimilarDFS(graph,src,dest,vis)) return false;
        }
        return true;
    }
}

File: C++/Graphs/21_sentenceSimilarity2.cpp
string find(string s, map<string, string> &parent)
{
    // check if s is present in map
    if (parent.find(s) == parent.end())
        parent[s] = s;

    if (parent[s] == s)
        return s;
    else
        return find(parent[s], parent);
}

bool areSentencesSimilarTwo(vector<string> &words1, vector<string> &words2, vector<vector<string>> &pairs)
{
    //
    map<string, string> parent;
    for (auto pair : pairs)
    {
        string s1 = find(pair[0], parent);
        string s2 = find(pair[1], parent);
        if (s1 != s2)
            parent[s1] = s2;
    }

    for (int i = 0; i < words1.size(); i++)
        if (find(words1[i], parent) != find(words2[i], parent))
            return false;

    return true;
}